src/OmniSharp.MiscellanousFiles/MiscellanousFiles.cs

```diff
@@ -0,0 +1,119 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using OmniSharp.FileSystem;
+using OmniSharp.FileWatching;
+using OmniSharp.Mef;
+using OmniSharp.Models.WorkspaceInformation;
+using OmniSharp.MSBuild;
+using OmniSharp.Services;
+
+namespace OmniSharp.MiscellanousFiles
+{
+    [ExtensionOrder(After = ProjectSystemNames.MSBuildProjectSystem)]
+    [ExportProjectSystem(ProjectSystemNames.MiscellanousFilesProjectSystem), Shared]
+    public class MiscellanousFilesProjectSystem : IProjectSystem
+    {
+        private readonly string miscFileExtension = ".cs";
+        public string Key { get; } = "MiscellanousFiles";
+        public string Language { get; } = LanguageNames.CSharp;
+        IEnumerable<string> IProjectSystem.Extensions => new[] { miscFileExtension };
+        public bool EnabledByDefault { get; } = true;
+
+        private readonly ConcurrentDictionary<string, DocumentId> _documents = new ConcurrentDictionary<string, DocumentId>();
+        private readonly OmniSharpWorkspace _workspace;
+        private readonly IFileSystemWatcher _fileSystemWatcher;
+        private readonly FileSystemHelper _fileSystemHelper;
+        private readonly ProjectSystem _projectSystem;
+        private readonly ILogger _logger;
+        private ProjectId _projectId;
+
+        [ImportingConstructor]
+        public MiscellanousFilesProjectSystem(OmniSharpWorkspace workspace, IFileSystemWatcher fileSystemWatcher, FileSystemHelper fileSystemHelper,
+            ILoggerFactory loggerFactory, [ImportMany] IEnumerable<Lazy<IProjectSystem, ProjectSystemMetadata>> projectSystems)
+        {
+            _workspace = workspace;
+            _fileSystemWatcher = fileSystemWatcher ?? throw new ArgumentNullException(nameof(fileSystemWatcher));
+            _fileSystemHelper = fileSystemHelper;
+            _logger = loggerFactory.CreateLogger<MiscellanousFilesProjectSystem>();
+            _projectSystem = (ProjectSystem)projectSystems.Where(ps => ps.Metadata.Name == ProjectSystemNames.MSBuildProjectSystem).First().Value;
+        }
+
+        Task<object> IProjectSystem.GetProjectModelAsync(string filePath)
+        {
+            return Task.FromResult<object>(null);
+        }
+
+        Task<object> IProjectSystem.GetWorkspaceModelAsync(WorkspaceInformationRequest request)
+        {
+            return Task.FromResult<object>(null);
+        }
+
+        void IProjectSystem.Initalize(IConfiguration configuration)
+        {
+            var allFiles = _fileSystemHelper.GetFiles("**/*.cs");
+            foreach (var filePath in allFiles)
+                AddIfMiscellanousFile(filePath);
+
+            _fileSystemWatcher.Watch(miscFileExtension, OnMiscellanousFileChanged);
+        }
+
+        private async void AddIfMiscellanousFile(string filePath)
+        {
+            //wait for the project system to get initialised
+            await _projectSystem.HasCompletedUpdateRequest();
```

Comment:
Shouldn't this be in the `Initialize` method rather than every miscellaneous file that gets added?

Reply:
After any file is added we need to wait for the MSBuild project system to signal that it has completed the update for that particular file. So, I think we need it here.